File: casbot2_py_demo/casbot2_py_demo/debug_joint_demo.py
"""Upper-body then whole-body hold commands; default PD configuration required."""
import time
import rclpy
from sensor_msgs.msg import JointState
from crb_ros_msg.msg import JointStateData, UpperJointData
from casbot2_tools.interface_cli import InterfaceDemo


class DebugJointDemo(InterfaceDemo):
    def __init__(self):
        super().__init__()
        self.latest = None
        self.create_subscription(JointState, '/joint_states', self.on_state, 10)

    def on_state(self, msg):
        self.latest = msg

    def run(self):
        deadline = time.monotonic() + 5.0
        while self.latest is None and time.monotonic() < deadline:
            rclpy.spin_once(self, timeout_sec=0.05)
        if self.latest is None or not self.latest.name:
            raise RuntimeError('缺少 /joint_states，不能构造保持姿态命令')
        names = list(self.latest.name)
        positions = list(self.latest.position)
        index = names.index('left_shoulder_pitch_joint')
        for service, msg, topic in (
            ('/motion/upper_body_debug', UpperJointData(vel_scale=0.05), '/upper_body_debug/joint_cmd'),
            ('/motion/whole_body_debug', JointStateData(), '/motion/debug/joint_cmd'),
        ):
            # Whole-body messages include every measured joint, retaining leg/arm pose.
            msg.name = [names[index]] if isinstance(msg, UpperJointData) else names
            msg.position = [positions[index]] if isinstance(msg, UpperJointData) else positions
            msg.velocity = [0.0] * len(msg.name)
            msg.effort = [0.0] * len(msg.name)
            pub = self.create_publisher(type(msg), topic, 10)
            self.cmd_switch_bool(service, True)
            try:
                for _ in range(20):
                    msg.header.stamp = self.get_clock().now().to_msg()
                    self.publish_once(pub, msg)
            finally:
                self.cmd_switch_bool(service, False)


def main():
    rclpy.init()
    node = DebugJointDemo()
    try:
        node.run()
        return 0
    except (RuntimeError, ValueError, TimeoutError) as exc:
        node.get_logger().error(str(exc))
        return 1
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == '__main__':
    raise SystemExit(main())

File: casbot2_cpp_demo/src/debug_joint_demo.cpp
#include <algorithm>
#include <chrono>
#include <memory>
#include <stdexcept>
#include <string>
#include "rclcpp/rclcpp.hpp"
#include "std_srvs/srv/set_bool.hpp"
#include "sensor_msgs/msg/joint_state.hpp"
#include "crb_ros_msg/msg/joint_state_data.hpp"
#include "crb_ros_msg/msg/upper_joint_data.hpp"
using namespace std::chrono_literals;

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = rclcpp::Node::make_shared("debug_joint_demo");
  sensor_msgs::msg::JointState::SharedPtr measured;
  auto sub = node->create_subscription<sensor_msgs::msg::JointState>(
    "/joint_states", 10, [&](sensor_msgs::msg::JointState::SharedPtr msg) { measured = msg; });
  auto switch_mode = [&](const std::string & service, bool enabled) {
    auto cli = node->create_client<std_srvs::srv::SetBool>(service);
    if (!cli->wait_for_service(3s)) { throw std::runtime_error(service + " unavailable"); }
    auto req = std::make_shared<std_srvs::srv::SetBool::Request>();
    req->data = enabled;
    auto future = cli->async_send_request(req);
    if (rclcpp::spin_until_future_complete(node, future, 3s) != rclcpp::FutureReturnCode::SUCCESS) {
      throw std::runtime_error(service + " timed out; transition state unknown");
    }
    auto response = future.get();
    if (!response->success) { throw std::runtime_error(response->message); }
  };
  std::string active_service;
  int result = 0;
  try {
    auto deadline = std::chrono::steady_clock::now() + 5s;
    while (rclcpp::ok() && !measured && std::chrono::steady_clock::now() < deadline) {
      rclcpp::spin_some(node);
      rclcpp::sleep_for(20ms);
    }
    if (!measured || measured->name.empty() || measured->name.size() != measured->position.size()) {
      throw std::runtime_error("No valid /joint_states snapshot");
    }
    const auto names = measured->name;
    const auto positions = measured->position;
    const auto found = std::find(names.begin(), names.end(), "left_shoulder_pitch_joint");
    if (found == names.end()) { throw std::runtime_error("Missing shoulder joint"); }
    const auto index = static_cast<size_t>(found - names.begin());
    auto upper_pub = node->create_publisher<crb_ros_msg::msg::UpperJointData>("/upper_body_debug/joint_cmd", 10);
    auto whole_pub = node->create_publisher<crb_ros_msg::msg::JointStateData>("/motion/debug/joint_cmd", 10);
    switch_mode("/motion/upper_body_debug", true);
    active_service = "/motion/upper_body_debug";
    crb_ros_msg::msg::UpperJointData upper;
    upper.vel_scale = 0.05f;
    upper.name = {names[index]};
    upper.position = {positions[index]};
    upper.velocity = {0.0}; upper.effort = {0.0};
    // Hold the measured pose. Empty PD requires use_default_kp_kd=true.
    for (int i = 0; i < 20 && rclcpp::ok(); ++i) {
      upper.header.stamp = node->now();
      upper_pub->publish(upper);
      rclcpp::spin_some(node); rclcpp::sleep_for(100ms);
    }
    switch_mode(active_service, false); active_service.clear();
    switch_mode("/motion/whole_body_debug", true);
    active_service = "/motion/whole_body_debug";
    crb_ros_msg::msg::JointStateData whole;
    whole.name = names; whole.position = positions;
    whole.velocity.assign(names.size(), 0.0); whole.effort.assign(names.size(), 0.0);
    for (int i = 0; i < 20 && rclcpp::ok(); ++i) {
      whole.header.stamp = node->now();
      whole_pub->publish(whole);
      rclcpp::spin_some(node); rclcpp::sleep_for(100ms);
    }
    switch_mode(active_service, false); active_service.clear();
  } catch (const std::exception & error) {
    RCLCPP_ERROR(node->get_logger(), "%s", error.what());
    result = 1;
  }
  if (!active_service.empty() && rclcpp::ok()) {
    try { switch_mode(active_service, false); }
    catch (const std::exception & error) { RCLCPP_ERROR(node->get_logger(), "%s", error.what()); }
  }
  rclcpp::shutdown();
  return result;
}
